### PR TITLE
fix: autoGrow prop on mobile is handled by native platform

### PR DESCRIPTION
**Pull request by Nodonisko**
### Explanation of Change
This PR is based on work from @hannojg in https://github.com/Expensify/App/pull/50238 and it's partially removing complicated `autoGrow` using layout events from `Text` to update dimension of `TextInput`. This is not necessary on native platforms because they can auto grow by default. 

Sadly on Android there is bug in RN that is causing component to be jumpy if I remove this autoGrow features that are adding some width buffer to Android component. For that reason I moved most of measuring logic into new component that is used only for Android + Web platforms so even more code is now reused across web and native.

### Fixed Issues

$ https://github.com/Expensify/App/issues/19961

### Tests

1. Sign in to any account
2. Go to any DM
3. Tap + in the compose box > Create Expense > Manual
4. Enter the amount
5. Verify that no jumpy behavior when entering the digits

### QA Steps

1. Sign in to any account
2. Go to any DM
3. Tap + in the compose box > Create Expense > Manual
4. Enter the amount
5. Verify that no jumpy behavior when entering the digits

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[adndroid-text-input.webm](https://github.com/user-attachments/assets/8d47b309-61b6-4531-8b91-5f11f5bd80de)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[android-web.webm](https://github.com/user-attachments/assets/ca461057-36bd-4163-9cea-4f14b36803d8)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/d21a0475-39c1-4c85-aa5b-12d0865d33b2

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/691a8917-3561-447b-b86d-7456686d9b6d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/d630d8e4-cddd-4a5c-b528-5bebff2be149

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/a12d490a-bddb-4765-8228-946c3ab3b8f7

</details>

### CODE REVIEW

**Comment by Nodonisko** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by hannojg** [comment]
Yes @Nodonisko is from margelo and took the task over from me 😊

**Comment by Nodonisko** [comment]
After some testing it looks that iOS is really fixed but now Android seems to glitch in very similar way. I will continue my investigation.

**Comment by mountiny** [comment]
@hoangzinh seems like Aimane is ooo so please take it over

**Comment by Nodonisko** [comment]
@hoangzinh Sorry I will update testing guide, but probably hold on with testing for now because I discovered issue on Android that needs to be fixed first https://github.com/Expensify/App/issues/19961#issuecomment-2755377925

**Comment by Nodonisko** [comment]
@hoangzinh PR is updated so hotfix is now used only for Android. I also reused more across web/native.

I have some issues with commit signing on new computer and will try to solve it tomorrow, but other than that it's ready for review.

**Comment by mountiny** [comment]
@hoangzinh what is your eta for the review? thakns!

File: src/components/TextInput/TextInputMeasurement/index.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                        styles.visibilityHidden,
+                    ]}
+                    onLayout={(e) => {
+                        if (e.nativeEvent.layout.width === 0 && e.nativeEvent.layout.height === 0) {
+                            return;
+                        }
+                        let additionalWidth = 0;
+                        if (isMobileSafari() || isSafari() || isMobileChrome()) {
+                            additionalWidth = 2;
+                        }
```
[the remark this answers is not in the document]
Comment (Nodonisko) [reply src/components/TextInput/TextInputMeasurement/index.tsx:75]: Yes, that should have been because of cursor at the end but even without that cursor is visible (ignore coloured background):

Screenshot 2025-04-14 at 21 08 35

**Comment by Nodonisko** [comment]
@hoangzinh I definitely noticed something when entering first number so I investigated it more and it look it's another bug (in same category "jumping").

I did some investigation and it seems that this is happening:

1. Component is mounted
2. Enter first digit after mount
3. Cursor stays before that digit
4. Enter second digit
5. Cursor jumps after seconds
6. Now cursor behaves correctly even if I delete everything and enter first digit again 

After some investigation it seems that is caused by using `selection` prop but not sure why.

You can see on screenshot that input grew correctly (highlighted by background) but cursor did not moved after digit. It just stays there until you write next digit.

Simulator Screenshot - iPhone 16 Pro - 2025-04-14 at 21 20 42

**Comment by Nodonisko** [comment]
@hoangzinh So we investigated more deeply and it seems to be another bug in RN related to selection/caret position which is controlled fully from JS in `AmountForm` which is quite uncommon. After some more investigation we noticed that **it is happening only when you use emulator** and HW keyboard of your computer so it shouldn't be big issue for now. If you use that onscreen keyboard it works fine. It is also happening on `main` branch so it is not introduced by this PR.

My proposal is to keep this PR as it is now.

Other options are:
1. Migrate `AmountForm` to recently introduced masked input, which should allow us to drop manual handling of `selection` which is causing issues
2. Fix it in RN but bug seems to be in code that is handling edge cases so it will be quite time consuming to test it

**Comment by Nodonisko** [comment]
@hoangzinh So I found simple fix in the end. 168e73feb2c74a322a8b178930ae70a8f1c72b4f

https://github.com/user-attachments/assets/af50b9bd-cb06-4d05-9ad9-2ca1b954d3c1

**Comment by mountiny** [comment]
I would not worry about issues that are only in dev with the hardware keyboard for this pr but lets fix it without a workaround

File: src/components/TextInput/BaseTextInput/implementation/index.native.tsx [hunk 2]
```diff
@@ -80,6 +81,10 @@ function BaseTextInput(
     }: BaseTextInputProps,
     ref: ForwardedRef<BaseTextInputRef>,
 ) {
+    // For iOS, we don't need to measure the text input because it already has auto grow behavior
+    // See TextInputMeasurement.ios.tsx for more details
+    const isExternalAutoGrowMeasurement = Platform.OS !== 'ios' && autoGrow;
```
[the remark this answers is not in the document]
Comment (Nodonisko) [reply src/components/TextInput/BaseTextInput/implementation/index.native.tsx:86]: Yea sure, so previous solution (still present for Android) measured size `Text` component and set same `width` to `TextInput` so width was specified with some absolute value.

Now on iOS we must not have any width specified or native auto grow won't work. So we don't want to apply anything that will set width manually like `flex1` or `width100`

File: src/components/TextInput/BaseTextInput/implementation/index.native.tsx [hunk 3]
```diff
@@ -360,8 +369,8 @@ function BaseTextInput(
                                 placeholderTextColor={placeholderTextColor ?? theme.placeholderText}
                                 underlineColorAndroid="transparent"
                                 style={[
-                                    styles.flex1,
-                                    styles.w100,
+                                    !autoGrow && styles.flex1,
```
[the remark this answers is not in the document]
Comment (Nodonisko) [reply src/components/TextInput/BaseTextInput/implementation/index.native.tsx:373]: Yea sure, so previous solution (still present for Android) measured size `Text` component and set same `width` to `TextInput` so width was specified with some absolute value.

Now on iOS we must not have any width specified or native auto grow won't work. So we don't want to apply anything that will set width manually like `flex1` or `width100`

**Comment by mountiny** [comment]
Sorry if I was not clear. We can use it for now, but we should not drop the ball and still try to pursue more complex but better fix (and removing the workaround then)

**Review by hoangzinh — APPROVED** [review]

**Review by mountiny — APPROVED** [review]
Looks great, thanks for your patience and congrats @Nodonisko for one of your first prs!

**Comment by aldo-expensify** [comment]
This PR seems to be causing this deploy blocker: https://github.com/Expensify/App/issues/60765

File: src/components/TextInput/TextInputMeasurement/index.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                 This text view is used to calculate width or height of the input value given textStyle in this component.
+                 This Text component is intentionally positioned out of the screen.
+             */}
+            {(!!autoGrow || !!autoGrowHeight) && !isAutoGrowHeightMarkdown && (
+                // Add +2 to width on Safari browsers so that text is not cut off due to the cursor or when changing the value
+                // Reference: https://github.com/Expensify/App/issues/8158, https://github.com/Expensify/App/issues/26628
+                // For mobile Chrome, ensure proper display of the text selection handle (blue bubble down).
+                // Reference: https://github.com/Expensify/App/issues/34921
+                <Text
+                    style={[
```
Comment (huult) [inline src/components/TextInput/TextInputMeasurement/index.tsx:63]: From https://github.com/Expensify/App/issues/76785, the displayed input and hidden measurement text don’t match due to styling differences, so their styles need to be synced
